scripts/jestPerTestSetup.ts

```diff
@@ -0,0 +1,75 @@
+import fs from 'fs-extra'
+import { resolve } from 'path'
+import { FastifyInstance } from 'fastify'
+import { Page } from 'playwright-chromium'
+
+export function slash(p: string): string {
+  return p.replace(/\\/g, '/')
+}
+
+// injected by the test env
+declare global {
+  // eslint-disable-next-line @typescript-eslint/no-namespace
+  namespace NodeJS {
+    interface Global {
+      page?: Page
+      fastifyRendererTestUrl?: string
+    }
+  }
+}
+
+let server: FastifyInstance
+let err: Error
+
+const logs: string[] = ((global as any).browserLogs = [])
+const onConsole = (msg) => { logs.push(msg.text()) }
+
+beforeAll(async () => {
+  const page = global.page
+  if (!page) {
+    return
+  }
+  try {
+    page.on('console', onConsole)
+
+    const testPath = expect.getState().testPath
+    const testName = slash(testPath).match(/test-apps\/([\w-]+)\//)?.[1]
+
+    // if this is a test placed under test-apps/xxx/test/
+    // start a vite server in that directory.
+    if (testName) {
+      const testAppsRoot = resolve(__dirname, '../packages/test-apps')
+      const srcDir = resolve(testAppsRoot, testName)
+
+      const serverEntrypoint = resolve(srcDir, 'server.ts')
+      if (!fs.existsSync(serverEntrypoint)) {
+        throw Error(`Missing server entrypoint file at: ${serverEntrypoint}`)
+      }
+
+      const { server: fastifyServer } = require(serverEntrypoint)
+      server = await fastifyServer()
+      await server.listen(3000)
+      const url = (global.fastifyRendererTestUrl = `http://localhost:${3000}`)
```

Comment:
Feels a little dirty to inject this via a global but I'm not aware of a better way! :shipit:

Reply:
You're right, I guess we could remove `fastifyRendererTestUrl` and just access the current page url using `page.url()` since we already navigated to it using `page.goto(url)` in this setup script.

Reply:
Removed `fastifyRendererTestUrl`